src/analysis/processing/qgsalgorithmtransect.cpp

```diff
@@ -0,0 +1,186 @@
+{
+  Side orientation = static_cast< QgsTransectAlgorithm::Side >( parameterAsInt( parameters, QStringLiteral( "SIDE" ), context ) );
+  double angle = fabs( parameterAsDouble( parameters, QStringLiteral( "ANGLE" ), context ) );
+  double length = parameterAsDouble( parameters, QStringLiteral( "LENGTH" ), context );
+
+  if ( orientation == QgsTransectAlgorithm::Both )
+    length /= 2.0;
+
+  std::unique_ptr< QgsFeatureSource > source( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !source )
+    return QVariantMap();
+
+  QgsFields fields = source->fields();
+
+  fields.append( QgsField( QStringLiteral( "TR_FID" ), QVariant::Int, QString(), 20 ) );
+  fields.append( QgsField( QStringLiteral( "TR_ID" ), QVariant::Int, QString(), 20 ) );
+  fields.append( QgsField( QStringLiteral( "TR_SEGMENT" ), QVariant::Int, QString(), 20 ) );
+  fields.append( QgsField( QStringLiteral( "TR_ANGLE" ), QVariant::Double, QString(), 5, 2 ) );
+  fields.append( QgsField( QStringLiteral( "TR_LENGTH" ), QVariant::Double, QString(), 20, 6 ) );
+  fields.append( QgsField( QStringLiteral( "TR_ORIENT" ), QVariant::Int, QString(), 1 ) );
+
+  QString dest;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, fields,
+                                          QgsWkbTypes::LineString, source->sourceCrs() ) );
+  if ( !sink )
+    return QVariantMap();
+
+  QgsFeatureIterator features = source->getFeatures( );
+
+  int current = -1;
+  int number = 0;
+  double step =  source->featureCount() > 0 ? 100.0 / source->featureCount() : 1;
+  QgsFeature feat;
+
+
+  while ( features.nextFeature( feat ) )
+  {
+    current++;
+    if ( feedback->isCanceled() )
+    {
+      break;
+    }
+
+    feedback->setProgress( current * step );
+    if ( !feat.hasGeometry() )
+      continue;
+
+    QgsGeometry inputGeometry = feat.geometry();
+
+    inputGeometry.convertToMultiType();
+    const QgsMultiLineString *multiLine = static_cast< QgsMultiLineString *  >( inputGeometry.geometry() );
+    for ( int id = 0; id < multiLine->numGeometries(); ++id )
+    {
+      const QgsLineString *line = static_cast< const QgsLineString * >( multiLine->geometryN( id ) );
+      QgsAbstractGeometry::vertex_iterator it = line->vertices_begin();
+      while ( it != line->vertices_end() )
+      {
+        QgsVertexId vertexId = it.vertexId();
+        int i = vertexId.vertex;
+        QgsFeature outFeat;
+        QgsAttributes attrs = feat.attributes();
+        attrs << current << number << i + 1 << angle <<
+              ( ( orientation == QgsTransectAlgorithm::Both ) ? length * 2 : length ) <<
+              orientation;
+        outFeat.setAttributes( attrs );
+        double angleAtVertex = line->vertexAngle( vertexId );
+        outFeat.setGeometry( calcTransect( line->pointN( i ), angleAtVertex, length, orientation, angle ) );
```

Comment:
Micro optimisation - you've already got the point here (from the iterator), so replacing `line->pointN( i )` with `*it` will avoid re-looking up the current vertex point.